pkg/restore/prioritize_group_version.go

```diff
@@ -0,0 +1,346 @@
+		}
+
+		// Priority 0: User Priority Version
+		if ugvs != nil {
+			uv := findSupportedUserVersion(ugvs[rg].Versions, tg.Versions, sg.Versions)
+			if uv != "" {
+				ctx.SetChosenGVToRestore(cgv, rg, sg.PreferredVersion.Version, uv)
+				ctx.log.Debugf("APIGroupVersionsFeatureFlag Priority 0: User defined API group version %s chosen for %s", uv, rg)
+				continue
+			}
+
+			ctx.log.Infof("Cannot find user defined version in both the cluster and backup cluster. Ignoring version %s for %s", uv, rg)
+		}
+
+		// Priority 1: Target Cluster Preferred Version
+		if versionsContain(sg.Versions, tg.PreferredVersion.Version) {
+			ctx.SetChosenGVToRestore(
+				cgv,
+				rg,
+				sg.PreferredVersion.Version,
+				tg.PreferredVersion.Version,
+			)
+			ctx.log.Debugf(
+				"APIGroupVersionsFeatureFlag Priority 1: Cluster preferred API group version %s found in backup for %s",
+				tg.PreferredVersion.Version,
+				rg,
+			)
+			continue
+		}
+
+		// Priority 2: Source Cluster Preferred Version
+		if versionsContain(tg.Versions, sg.PreferredVersion.Version) {
+			ctx.SetChosenGVToRestore(
+				cgv,
+				rg,
+				sg.PreferredVersion.Version,
+				sg.PreferredVersion.Version,
+			)
+			ctx.log.Debugf(
+				"APIGroupVersionsFeatureFlag Priority 2: Cluster preferred API group version not found in backup. Using backup preferred version %s for %s",
+				sg.PreferredVersion.Version,
+				rg,
+			)
+			continue
+		}
+
+		// Priority 3: The Common Supported Version with the Highest Kubernetes Version Priority
+		for _, tv := range tg.Versions[1:] {
+			if versionsContain(sg.Versions[1:], tv.Version) {
+				ctx.SetChosenGVToRestore(cgv, rg, sg.PreferredVersion.Version, tv.Version)
+				continue OUTER
+			}
+			ctx.log.Debugf(
+				"APIGroupVersionsFeatureFlag Priority 3: Common supported but not preferred API group version %s chosen for %s",
+				tv.Version,
+				rg,
+			)
+		}
+
+		// Use default group version.
+		ctx.SetChosenGVToRestore(cgv, rg, "", "")
+		ctx.log.Debugf(
+			"APIGroupVersionsFeatureFlag: Unable to find supported priority API group version. Using backup preferred version %s for %s (default behavior without feature flag).",
+			tg.PreferredVersion.Version,
+			rg,
+		)
+	}
+
+	return nil
+}
+
+// gatherSTUVersions collects the source, target, and user priority versions.
+func (ctx *restoreContext) gatherSTUVersions() (
+	map[string]metav1.APIGroup,
+	[]metav1.APIGroup,
+	map[string]metav1.APIGroup,
+	error,
+) {
+	sourceRGVersions, err := archive.NewParser(ctx.log, ctx.fileSystem).ParseGroupVersions(ctx.restoreDir)
+	if err != nil {
+		return nil, nil, nil, errors.Wrap(err, "parsing versions from directory names")
+	}
+
+	// Sort the versions in the APIGroups in sourceRGVersions map values.
+	for _, src := range sourceRGVersions {
+		k8sPrioritySort(src.Versions)
+	}
+
+	targetGroupVersions := ctx.discoveryHelper.APIGroups()
+
+	// Sort the versions in the APIGroups slice in targetGroupVersions.
+	for _, tar := range targetGroupVersions {
```

Comment:
When I saw the variable named "tar", I thought it was from the tarball. suggest renaming/expanding the same of the variable to "target" or something more descriptive.

Reply:
Will do.